packages/actor-query-operation-from-quad/lib/ActorQueryOperationFromQuad.ts

```diff
@@ -16,58 +16,124 @@ export class ActorQueryOperationFromQuad extends ActorQueryOperationTypedMediate
+    if (operation.type === 'pattern' || operation.type === 'path') {
+      const patternGraph: RDF.Term = operation.graph;
+      const defaultGraph: boolean = patternGraph.termType === 'DefaultGraph';
+      if (defaultGraphs.length === 1) {
+        if (defaultGraph) {
+          const graph: RDF.Term = defaultGraphs[0];
+          // If the pattern's graph is the default graph, replace the graph directly.
+          if (operation.type === 'pattern') {
+            return ActorQueryOperationFromQuad.FACTORY
+              .createPattern(operation.subject, operation.predicate, operation.object, graph);
+          } else {
+            return ActorQueryOperationFromQuad.FACTORY
+              .createPath(operation.subject, operation.predicate, operation.object, graph);
+          }
+        } else {
+          // Otherwise, keep the non-default graph as-is.
+          return operation;
+        }
       } else {
-        return defaultGraph ? ActorQueryOperationFromQuad.unionOperations(graphs.map(
-          (graph: RDF.Term) => ActorQueryOperationFromQuad.applyOperationGraph(operation, [graph])))
+        return defaultGraph ? ActorQueryOperationFromQuad.unionOperations(defaultGraphs.map(
+          (graph: RDF.Term) => ActorQueryOperationFromQuad.applyOperationDefaultGraph(operation, [graph])))
           : operation;
       }
     }
 
-    // Otherwise, copy the operation and recursively call this method.
-    const copiedOperation: Algebra.Operation = <any> {};
-    for (const key of Object.keys(operation)) {
-      if (Array.isArray(operation[key])) {
-        if (key === 'variables') {
-          copiedOperation[key] = operation[key];
+    return ActorQueryOperationFromQuad.copyOperation(operation,
+      (subOperation: Algebra.Operation) => this.applyOperationDefaultGraph(subOperation, defaultGraphs));
+  }
+
+  /**
+   * Recursively transform the given operation to use the given graphs as named graph
+   * This will (possibly) create a new operation and not modify the given operation.
+   * @param {Operation} operation An operation.
+   * @param {RDF.Term[]} namedGraphs Graph terms.
+   * @return {Operation} A new operation.
+   */
+  public static applyOperationNamedGraph(operation: Algebra.Operation, namedGraphs: RDF.Term[],
+                                         defaultGraphs: RDF.Term[]): Algebra.Operation {
+    // If the operation is a Pattern or Path, change the graph.
+    if (operation.type === 'pattern' || operation.type === 'path') {
+      const patternGraph: RDF.Term = operation.graph;
+      if (patternGraph.termType === 'DefaultGraph') {
+        // SPARQL spec (8.2) describes that when FROM NAMED's are used without a FROM, the default graph must be empty.
+        // The FROMs are transformed before this step to a named node, so this will not apply to this case anymore.
+        return { type: 'bgp', patterns: [] };
+      } else if (patternGraph.termType === 'Variable') {
```

Comment:
The problem is that we can't use VALUES for binding blank nodes.

I actually think it might be best if we convert all blank nodes in a query to (unique) variables. It would avoid us having to type _if variable or blank node_ all the time. I'll open an issue if you agree.

Reply:
Yea I think we can do that. In comunica though, not in sparqlalgebra.js itself.